src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleEqualityTests.cs

```diff
@@ -0,0 +1,2814 @@
+
+        public override bool Equals(object o)
+            => throw null;
+
+        public override int GetHashCode()
+            => throw null;
+    }
+}
+public class C
+{
+    public static void Main()
+    {
+        var t1 = (1, 1);
+        var t2 = (2, 2);
+        if (t1 == t2)
+        {
+            return;
+        }
+    }
+}
+";
+            var comp = CreateStandardCompilation(source, options: TestOptions.DebugExe);
+            comp.VerifyDiagnostics();
+            CompileAndVerify(comp, expectedOutput: "Correct operator was called");
+        }
+
+        [Fact]
+        public void TestNaN()
+        {
+            var source = @"
+public class C
+{
+    public static void Main()
+    {
+        var t1 = (System.Double.NaN, 1);
+        var t2 = (System.Double.NaN, 1);
+        System.Console.Write($""{t1 == t2} {t1.Equals(t2)} {t1 != t2} {t1 == (System.Double.NaN, 1)}"");
+    }
+}
+";
+            var comp = CreateStandardCompilation(source, references: s_valueTupleRefs, options: TestOptions.DebugExe);
+            comp.VerifyDiagnostics();
+            CompileAndVerify(comp, expectedOutput: "False True True False");
+        }
+
+        [Fact]
+        public void TestTopLevelDynamic()
+        {
+            var source = @"
+public class C
+{
+    public static void Main()
+    {
+        dynamic d1 = (1, 1);
+
+        try
+        {
+            _ = d1 == (1, 1);
+        }
+        catch (Microsoft.CSharp.RuntimeBinder.RuntimeBinderException e)
+        {
+            System.Console.WriteLine(e.Message);
+        }
+
+        try
+        {
+            _ = d1 != (1, 2);
+        }
+        catch (Microsoft.CSharp.RuntimeBinder.RuntimeBinderException e)
+        {
+            System.Console.WriteLine(e.Message);
+        }
+    }
+}
+";
+            var comp = CreateStandardCompilation(source, options: TestOptions.DebugExe,
+                references: new[] { ValueTupleRef, SystemRuntimeFacadeRef, CSharpRef, SystemCoreRef });
+            comp.VerifyDiagnostics();
+
+            CompileAndVerify(comp, expectedOutput:
+@"Operator '==' cannot be applied to operands of type 'System.ValueTuple<int,int>' and 'System.ValueTuple<int,int>'
+Operator '!=' cannot be applied to operands of type 'System.ValueTuple<int,int>' and 'System.ValueTuple<int,int>'");
+        }
+
+        [Fact]
+        public void TestComparisonWithDeconstruction()
+        {
+            var source = @"
+public class C
+{
+    public static void Main()
+    {
+        var b1 = (1, 2) == ((_, _) = new C());
+        var b2 = (1, 42) != ((_, _) = new C());
+        var b3 = (1, 42) == ((_, _) = new C()); // false
+        var b4 = ((_, _) = new C()) == (1, 2);
+        var b5 = ((_, _) = new C()) != (1, 42);
+        var b6 = ((_, _) = new C()) == (1, 42); // false
+        System.Console.Write($""{b1} {b2} {b3} {b4} {b5} {b6}"");
+    }
+    public void Deconstruct(out int x, out int y)
+    {
+        x = 1;
+        y = 2;
+    }
+}
```

Comment:
Is comparison supported directly with an instance of the type that supports deconstruction: `(1, 2) == new C()`? Perhaps add a test.

Reply:
That's not supported. Added a note to speclet and added a test.

---
In reply to: [167993979](https://github.com/dotnet/roslyn/pull/23356#discussion_r167993979) [](ancestors = 167993979)